packages/financial-templates-lib/test/proxy-transaction-handler/DSProxyManager.js

```diff
@@ -0,0 +1,159 @@
+    // should throw with invalid addresses
+    assert.throws(() => {
+      new DSProxyManager({
+        logger: spyLogger,
+        web3,
+        gasEstimator,
+        account: "not an address",
+        dsProxyFactoryAddress: dsProxyFactory.address,
+        dsProxyFactoryAbi: DSProxyFactory.abi,
+        dsProxyAbi: DSProxy.abi
+      });
+    });
+  });
+
+  it("Initialization correctly deploys a DSProxy if the user has not deployed one", async function() {
+    await dsProxyManager.initializeDSProxy();
+
+    dsProxyFactory = await DSProxyFactory.at(dsProxyManager.getDSProxyFactoryAddress());
+
+    // Seeing the EOA had no DSProxy before, it should have deployed a new DSProxy with the contractCreator set as owner.
+    const logs = await dsProxyFactory.getPastEvents("Created", { fromBlock: 0 });
+
+    assert.equal(logs.length, 1);
+    assert.equal(logs[0].returnValues.owner, contractCreator);
+    assert.equal(await (await DSProxy.at(logs[0].returnValues.proxy)).owner(), contractCreator);
+
+    assert.isTrue(spyLogIncludes(spy, -1, "DSProxy has been deployed for the EOA"));
+    assert.isTrue(spyLogIncludes(spy, -1, contractCreator));
+    assert.isTrue(spyLogIncludes(spy, -1, logs[0].returnValues.proxy));
+  });
+
+  it("Initialization correctly detects an existing DSProxy if the user has one already", async function() {
+    const createDSProxyTx = await dsProxyFactory.build();
+    await dsProxyManager.initializeDSProxy();
+    assert.isTrue(spyLogIncludes(spy, -1, "DSProxy has been loaded in for the EOA"));
+    assert.isTrue(spyLogIncludes(spy, -1, contractCreator)); // The EOA should be included.
+    assert.isTrue(spyLogIncludes(spy, -1, createDSProxyTx.logs[0].args.proxy)); // The log should contain our previous DSProxy.
+  });
+
+  it("Can send functions to DSProxy using an existing deployed library", async function() {
+    await dsProxyManager.initializeDSProxy();
+    const dsProxyAddress = dsProxyManager.getDSProxyAddress();
+
+    await testToken.mint(dsProxyAddress, toWei("1000"), { from: contractCreator });
+
+    // Deploy a tokenSender contract containing library code for the tx. the TokenSender contract enables a DSProxy to send
+    // any tokens that it might have within it's wallet. It is used to showcase how a more complex transaction would occur
+    // with the DSProxy acting in place of an EOA.
+    tokenSender = await TokenSender.new();
+
+    // Encode the send transaction. This is what will be executed within the DSProxy.
+    const sendTokenTx = tokenSender.contract.methods
+      .transferERC20(testToken.address, contractCreator, toWei("10"))
+      .encodeABI();
+
+    // Call the method on the tokenSender(library) with the sendTokenTxData. Note this is called from the perspective
+    // of the DSProxy and should send some of the DSProxy tokens.
+    const dsProxyCallReturn = await dsProxyManager.callFunctionOnExistingLibrary(tokenSender.address, sendTokenTx);
```

Comment:
this is really cool to see.  i dont really fully understand how this is possible. The ds proxy does not need to know anything about the token library other than address? I guess the sendTokenTx has all the info it needs to execute a call on the contract?

Reply:
yup! the dsProxy just needs to know where the code is deployed (the `tokenSender.address`) and then it can execute the transaction  `sendTokenTx` on the library. Take a look at the DSProxy smart contract code and see the `execute` method. This function either runs a function on a existing library (in the case of this call) or if overloaded runs on a contract that is first deployed. 

The key part of the `execute` method to look at is the assembly that uses the `delegatecall` opcode. this runs the provided transaction data (in this case `sendTokenTx`) in the context of the contract (the DSProxy) and executes the code at the location specified in the address.